bin/ncp/SECURITY/multi-factor-authentication.sh

```diff
@@ -0,0 +1,286 @@
+  local auth_method=""
+
+  [[ "$enable_pubkey_only" == "yes" ]] && auth_method="publickey"
+  [[ "$enable_pw_only" == "yes" ]] && auth_method="${auth_method} password"
+
+  [[ "$enable_totp_and_pw" == "yes" ]] && auth_method="keyboard-interactive"
+  [[ "$enable_pubkey_and_pw" == "yes" ]] && auth_method="${auth_method} publickey,password"
+
+  patch_pam_ssh_config || return 2
+  patch_sshd_config "$auth_method" || return 1
+}
+
+setup_totp_secret() {
+  local ssh_user="${1?}"
+  local ssh_user_home="${2?}"
+
+  [[ "$reset_totp_secret" == "yes" ]] \
+  && [[ -f "$ssh_user_home/.google_authenticator" ]] \
+  && {
+    echo "Deleting google authenticator configuration"
+    su "$ssh_user" -c "chmod u+w '${ssh_user_home}/.google_authenticator'"
+    su "$ssh_user" -c "rm '${ssh_user_home}/.google_authenticator'"
+  }
+
+
+  if [[ "$enable_totp_and_pw" == "yes" ]] && [[ ! -f "${ssh_user_home}/.google_authenticator" ]]
+  then
+    echo "We will now generate TOTP a client secret for your ssh user ('$ssh_user')."
+    echo "Please store the following information in a safe place. Use your secret key or scan the QR code (terminal only) to setup your authenticator app."
+    echo ""
+    echo "<pre style=\"font-family: mono;\">"
+    su "$ssh_user" -c "google-authenticator -tdf -Q UTF8 -w 1 --no-rate-limit"
+    echo "</pre>"
+  fi
+}
+
+restore() {
+  local ret=0
+  patch_pam_ssh_config --reset
+  ret=$((ret + $?))
+  patch_sshd_config --reset
+  ret=$((ret + $?))
+  return $ret
+}
+
+check_configuration_is_valid() {
+    if { [[ "$enable_pubkey_and_pw" == "yes" ]] || [[ "$enable_pubkey_only" == "yes" ]]; } && [[ -z "$SSH_PUBLIC_KEY" ]]
+    then
+      echo "ERROR: Public key reliant authentication methods have been enabled, but no publick key has been specified. Aborting..."
+      return 1
+    fi
+
+    if [[ "$enable_totp_and_pw" == "yes" ]] && [[ "$enable_pubkey_and_pw" == "yes" ]]
+    then
+      echo "Due to limitations of the sshd configuration, totp+pw and other authentication methods involving a password can't be enabled at the same time."
+      echo "Aborting..."
+      return 1
+    fi
+}
+
+################################################################
+
+cleanup() {
+  restore
+  [[ -d "${PAMD_BACKUP_PATH}" ]] && rm -r "${PAMD_BACKUP_PATH}"
+}
+
+install() {
+  apt install -y libpam-google-authenticator
+}
+
+is_active() {
+  grep -q -e "AuthenticationMethods.*keyboard-interactive" -e "AuthenticationMethods.*publickey" "${SSHD_CONFIG_PATH}" \
+  || grep -q -e "sshd-mfa" "${PAMD_PATH}/sshd"
+}
+
+configure() {
+
+  local active enable_totp_and_pw enable_pubkey_and_pw enable_pubkey_only enable_pw_only reset_totp_secret ssh_pubkey
+  enable_totp_and_pw="$ENABLE_TOTP_AND_PASSWORD"
+  enable_pubkey_and_pw="$ENABLE_PUBLIC_KEY_AND_PASSWORD"
+  enable_pubkey_only="$ENABLE_PUBLIC_KEY_ONLY"
+  enable_pw_only="$ENABLE_PASSWORD_ONLY"
+  reset_totp_secret="$RESET_TOTP_SECRET"
+  active="yes"
+  ssh_pubkey="$(unescape "$SSH_PUBLIC_KEY")"
```

Comment:
if we want to allow spaces, it sounds like we should use bash arrays

Reply:
That doesn't sound right to me. IMO, bash arrays are rather unintuitive for this purpose, because the array fields have no meaning (it's not a space separated set of strings, but a string containing spaces).

Reply:
Why would you prefer bash arrays here?

Reply:
what is the issue exactly? is it the sanitizing issue above? if the issue is ncp-config getting confused by spaces we would have to see if it makes sense to use arrays. It is probably a bad idea in terms of sanitizing/security to allow spaces as an input but we would have to take a close look

Reply:
The issue is that many values for ncp apps might validly contain spaces (e. g. paths or, like in this case, an ssh pubkey).
Therefore we need a way to provide unsanitized input values.
This might make sense for more than spaces (although not required in this case).

Have you had a look at my change to ncp-launcher.php?

Reply:
The problem with allowing spaces (in any form) is that an attacker could potentially craft a command with arguments if he is able to bypass other restrictions.

If we disallow spaces, at least all the attacker can do at that point is to run a command with arguments which would hopefully be much less useful for him.

How about uploading the pubkey, or having a special field that saves the unsanitized value to a file?

Reply:
That kind of attack is not possible if unsafe values are handled the way I implemented it.

- Spaces are being replaced by "%SPACE%" in the input values.
- They must be explicitly unescaped in bash in order to get the 'unsafe' value

If you are aware that a value might contain spaces, you can easily avoid that kind of argument injection.

In this particular case, a file upload would work as well - however, people are complaining about missing support for spaces (e. g. in paths) since forever, so that would solve a lot more cases than just ssh public key provision.

Reply:
Mmm I don't see a way where spaces would be an issue in this case. Makes me a bit uneasy, but if we are extra careful looks ok. Hopefully we are not missing anything.

Reply:
Well, if you prefer another option - I can work with any of the mentioned ideas for this use case :)
But personally, I'd prefer this implementation (my second pick would be bash arrays, since it provides the same functionality but is harder to implement and less intuitive).

Reply:
like I said, I don't see how in this case your implementation could be insecure. I think it is fine if you also triple check that there is no way to inject commands.

Reply:
Alright. :) 

I was just saying that, because my implementation includes changes to ncp-launcher.php and library.sh, meaning it will be available to other apps in the future.

Reply:
great, let me know when you are happy with this and I'll give it a deeper review

Reply:
I am now :)